File: jmlparser-jml-tools/src/main/java/com/github/jmlparser/smt/solver/Solver.java
package com.github.jmlparser.smt.solver;

import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.io.Reader;
import java.util.concurrent.ForkJoinPool;
import java.util.concurrent.ForkJoinTask;

/**
 * @author Alexander Weigl
 * @version 1 (08.08.22)
 */
public class Solver {
    public ForkJoinTask<SolverAnswer> runAsync(String input) {
        return ForkJoinPool.commonPool().submit(() -> run(input));
    }

    public SolverAnswer run(String input) throws IOException {
        return run(writer -> writer.println(input));
    }

    protected Process startSmtSolver() throws IOException {
        ProcessBuilder pb = new ProcessBuilder("sh", "-c", "z3 -smt2 -in");
        return pb.start();
    }

    public SolverAnswer run(AppendableTo query) throws IOException {
        Process p = startSmtSolver();
        try (PrintWriter out = new PrintWriter(p.getOutputStream());
             Reader in = new InputStreamReader(p.getInputStream())) {
            query.appendTo(out);
            out.close();
            return new SolverAnswer(SExprParser.parseAll(in));
        } finally {
            p.destroyForcibly();
        }
    }
}

File: tools/smt/src/main/kotlin/io/github/jmltoolkit/smt/solver/Solver.kt
package io.github.jmltoolkit.smt.solver

import java.io.IOException
import java.io.InputStreamReader
import java.io.PrintWriter
import java.util.concurrent.ForkJoinPool
import java.util.concurrent.ForkJoinTask

/**
 * @author Alexander Weigl
 * @version 1 (08.08.22)
 */
class Solver {
    fun runAsync(input: String): ForkJoinTask<SolverAnswer> {
        return ForkJoinPool.commonPool().submit<SolverAnswer> { run(input) }
    }

    @Throws(IOException::class)
    fun run(input: String): SolverAnswer {
        return run { writer: PrintWriter -> writer.println(input) }
    }

    @Throws(IOException::class)
    protected fun startSmtSolver(): Process {
        val pb = ProcessBuilder("sh", "-c", "z3 -smt2 -in")
        return pb.start()
    }

    @Throws(IOException::class)
    fun run(query: AppendableTo): SolverAnswer {
        val p = startSmtSolver()
        try {
            PrintWriter(p.outputStream).use { out ->
                InputStreamReader(p.inputStream).use {
                    query.appendTo(out)
                    out.close()
                    return SolverAnswer(SExprParser.parseAll(it))
                }
            }
        } finally {
            p.destroyForcibly()
        }
    }
}
